lib/TPP/Transforms/HoistVectorTransfers.cpp

```diff
@@ -0,0 +1,198 @@
+                return rewriter.notifyMatchFailure(contractOp, "Not a linalg tile + vector contract operation");
+
+        auto oldNForOp = oldReductionForOp->getParentOfType<scf::ForOp>();
+        if (!oldNForOp)
+                return rewriter.notifyMatchFailure(contractOp, "Not a linalg tile + vector contract operation");
+
+        auto oldMForOp = oldNForOp->getParentOfType<scf::ForOp>();
+        if (!oldMForOp)
+                return rewriter.notifyMatchFailure(contractOp, "Not a linalg tile + vector contract operation");
+
+
+        // Check the vector contract operation satisfies the required pattern.
+        // Check the ACC, LHS, and RHS of contract operation
+        auto vectorReadOpAcc = contractOp.getAcc().getDefiningOp<mlir::vector::TransferReadOp>();
+        if (!vectorReadOpAcc)
+                return failure();
+
+        auto subviewOp = vectorReadOpAcc.getOperand(0).getDefiningOp<memref::SubViewOp>();
+        if (!subviewOp)
+                return failure();
+
+        auto vectorReadOpLHS = contractOp.getLhs().getDefiningOp<mlir::vector::TransferReadOp>();
+        if (!vectorReadOpLHS)
+                return failure();
+
+        auto vectorReadOpRHS = contractOp.getRhs().getDefiningOp<mlir::vector::TransferReadOp>();
+        if (!vectorReadOpRHS)
+                return failure();
+
+
+
+        // Check the operation type MatMul, B-MatMul, or BR-MatMul
+        auto contractIteratorTypes = contractOp.getIteratorTypesArray();
+        MatMulType matmulType;
+
+        if (contractIteratorTypes.size() > 3) {
+                matmulType = contractIteratorTypes[contractIteratorTypes.size() - 4] == vector::IteratorType::parallel ? MatMulType::Batch : MatMulType::BatchReduce;
+                if (matmulType == MatMulType::Batch)
+                        return rewriter.notifyMatchFailure(contractOp, "Batch matmul not supported");
+        } else if (contractIteratorTypes.size() == 3) {
+                matmulType = MatMulType::Standard;
+        } else {
+                return rewriter.notifyMatchFailure(contractOp, "The vector contract operation is not a gemm");
+        }
+
+        auto vectorReadOpLHSType = cast<ShapedType>(vectorReadOpLHS.getType());
+        auto vectorReadOpRHSType = cast<ShapedType>(vectorReadOpRHS.getType());
+
+        if (matmulType == MatMulType::BatchReduce && (vectorReadOpLHSType.getRank() != 3 || vectorReadOpRHSType.getRank() != 3))
+                return failure();
+
+        if (matmulType == MatMulType::Standard  && (vectorReadOpLHSType.getRank() != 2 || vectorReadOpRHSType.getRank() != 2))
+                return failure();
+
+        //Check the K-dim to be 1
+        int64_t K = vectorReadOpLHSType.getDimSize(vectorReadOpLHSType.getRank() - 1);
+        if (K != 1)
+                return rewriter.notifyMatchFailure(contractOp, "K dim is not 1");
+
+
+        // Move the vector transfer read before the reduction and k loop
+        rewriter.setInsertionPointAfter(subviewOp);
```

Comment:
That's dangerous. You want to move it relative to the loop region, not some arbitrary operation in there.

Reply:
The `subviewOp` is the operand of `vectorTransferRead`. So, the `setInsertionPointAfter` is set on `subviewOp` to move the  `vectorTransferReadOp` of `C`. But, it's okay now the `setInsertionPoint` is set to the `Reduction` loop.

Reply:
Oh, so if the subview is still a producer of the transfer read's argument, then your original line was not wrong, but incomplete. We need to make sure that you add it to the right level AND after all its producers.